streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,326 @@
+
+    Return:
+        mds_path: actual local and remote path were used
+        fail_count: number of records failed to be converted
+    Raises:
+        ValueError: If dataframe is not provided
+
+    Note:
+        - The method creates a SparkSession if not already available.
+        - If 'sample_ratio' is provided, the input data will be randomly sampled.
+        - The 'ppfn_kwargs' dictionaries can be used to pass additional
+          keyword arguments to the udf_iterable.
+    """
+
+    def write_mds(iterator: Iterable):
+
+        context = TaskContext.get()
+        if context is not None:
+            id = context.taskAttemptId()
+        else:
+            raise Exception('TaskContext.get() returns None')
+
+        if isinstance(mds_path, str):  # local
+            output = os.path.join(mds_path, f'{id}')
+            out_file_path = output
+        else:
+            output = (os.path.join(mds_path[0], f'{id}'), os.path.join(mds_path[1], f'{id}'))
+            out_file_path = output[0]
+
+        if mds_kwargs:
+            kwargs = mds_kwargs.copy()
+            kwargs['out'] = output
+        else:
+            kwargs = {}
+
+        if merge_index:
+            kwargs['keep_local'] = True  # need to keep local to do merge
+
+        count = 0
+
+        with MDSWriter(**kwargs) as mds_writer:
+            for pdf in iterator:
+                if udf_iterable is not None:
+                    records = udf_iterable(pdf, **udf_kwargs or {})
+                else:
+                    records = pdf.to_dict('records')
+                assert is_iterable(
+                    records
+                ), f'pandas_processing_fn needs to return an iterable instead of a {type(records)}'
+
+                for sample in records:
+                    try:
+                        mds_writer.write(sample)
+                    except:
+                        logger.warning(f'failed to write sample: {sample}')
+                        count += 1
+                        raise Exception(f'failed to write sample: {sample}')
+
+        yield pd.concat(
+            [pd.Series([out_file_path], name='mds_path'),
+             pd.Series([count], name='fail_count')],
+            axis=1)
+
+    if dataframe is None or dataframe.count() == 0:
+        raise ValueError(f'input dataframe is none or empty!')
+
+    if not mds_kwargs:
+        mds_kwargs = collections.defaultdict(str)
+
+    if not udf_kwargs:
+        udf_kwargs = collections.defaultdict(str)
+
+    if 'out' not in mds_kwargs:
+        raise ValueError(f'out and columns need to be specified in mds_kwargs')
+
+    if 'columns' not in mds_kwargs:
+        logger.warning(
+            "User's discretion required: columns arg is missing from mds_kwargs. Will be auto inferred"
+        )
+        mds_kwargs['columns'] = infer_dataframe_schema(dataframe)
+    else:
+        infer_dataframe_schema(dataframe, mds_kwargs['columns'])
+
+    if 0 < sample_ratio < 1:
```

Comment:
Why do need `sample_ratio` ? Do you think it should be part of this method or let user provide the final dataframe that they wanted to convert to MDS ?

Reply:
Cody mentioned this may be useful for non-experienced spark users. And it was handy during debugging. So I would rather keep this option for the users.

Reply:
@codestar12 Do you think the `dataframeToMDS()` method should support `sample_ratio` ? Or we can expect user to provide a final dataframe ?

What is the use-case of `sample_ratio` here ?

Reply:
I wouldn't mind a sample_ratio. I think @zankner has found it can be non-trival sometimes to properly sample with spark in ways that map to the mental model of how I would expect it to in streaming

Reply:
I removed sample_ratio option.